flink-runtime/src/test/java/org/apache/flink/runtime/rest/handler/legacy/ExecutionGraphCacheTest.java

```diff
@@ -0,0 +1,357 @@
+		final AccessExecutionGraph accessExecutionGraph2 = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		when(jobManagerGateway.requestJob(eq(jobId1), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph1));
+		when(jobManagerGateway.requestJob(eq(jobId2), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph2));
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+
+			CompletableFuture<AccessExecutionGraph> executionGraph1Future = executionGraphCache.getExecutionGraph(jobId1, jobManagerGateway);
+
+			CompletableFuture<AccessExecutionGraph> executionGraph2Future = executionGraphCache.getExecutionGraph(jobId2, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph1, executionGraph1Future.get());
+
+			assertEquals(accessExecutionGraph2, executionGraph2Future.get());
+
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId1), any(Time.class));
+			verify(jobManagerGateway, times(1)).requestJob(eq(jobId2), any(Time.class));
+
+			Thread.sleep(timeToLive.toMilliseconds());
+
+			executionGraphCache.cleanup();
+
+			CompletableFuture<AccessExecutionGraph> executionGraph1Future2 = executionGraphCache.getExecutionGraph(jobId1, jobManagerGateway);
+
+			CompletableFuture<AccessExecutionGraph> executionGraph2Future2 = executionGraphCache.getExecutionGraph(jobId2, jobManagerGateway);
+
+			assertEquals(accessExecutionGraph1, executionGraph1Future2.get());
+
+			assertEquals(accessExecutionGraph2, executionGraph2Future2.get());
+
+			verify(jobManagerGateway, times(2)).requestJob(eq(jobId1), any(Time.class));
+			verify(jobManagerGateway, times(2)).requestJob(eq(jobId2), any(Time.class));
+		}
+	}
+
+	/**
+	 * Tests that concurrent accesses only trigger a single AccessExecutionGraph request.
+	 */
+	@Test
+	public void testConcurrentAccess() throws Exception {
+		final Time timeout = Time.milliseconds(100L);
+		final Time timeToLive = Time.hours(1L);
+		final JobID jobId = new JobID();
+
+		final AccessExecutionGraph accessExecutionGraph = mock(AccessExecutionGraph.class);
+
+		final JobManagerGateway jobManagerGateway = mock(JobManagerGateway.class);
+		when(jobManagerGateway.requestJob(eq(jobId), any(Time.class))).thenReturn(CompletableFuture.completedFuture(accessExecutionGraph));
+
+		final int numConcurrentAccesses = 10;
+
+		final ArrayList<CompletableFuture<AccessExecutionGraph>> executionGraphFutures = new ArrayList<>(numConcurrentAccesses);
+
+		try (ExecutionGraphCache executionGraphCache = new ExecutionGraphCache(timeout, timeToLive)) {
+			for (int i = 0; i < numConcurrentAccesses; i++) {
+				CompletableFuture<AccessExecutionGraph> executionGraphFuture = CompletableFuture
+					.supplyAsync(
+						() -> executionGraphCache.getExecutionGraph(jobId, jobManagerGateway),
+						TestingUtils.defaultExecutor())
```

Comment:
isn't the default executor single-threaded?

Reply:
This is a good point. Will change it.